libflux: add flux_buffer_write_watcher_close()

grondo opened the pull request:
As discussed in #1546, add a simple method for scheduling a call to `close(2)` once buffer has been emptied for flux buffer write watchers. This is basically the simplest approach for now while we continue to discuss alternatives.

Also added is the `reactorcat` test case, which implements something like `cat(1)` with the Flux reactor and buffer watchers.

Still need to add unit tests and a sharness test driver for `reactorcat`.

Fixes #1546

garlick commented:
Just scanned through this and LGTM.

chu11 commented on src/common/libflux/reactor.c:
```diff
@@ -563,6 +563,26 @@ flux_buffer_t *flux_buffer_write_watcher_get_buffer (flux_watcher_t *w)
+{
+    if (w) {
+        struct ev_buffer_write *evw = w->data;
+        evw->eof = true;
```
Should we add  ```if (!evw->eof && !evw->closed)``` so users can't close multiple times?

grondo replied:
I thought it should be idempotent. However, you are right might as well check for evw->closed otherwise eof is reset and the close callback could be called again. 

Should the function therefore have a return value and return EINVAL if `evw->closed` and EINPROGRESS if `evw->eof && !evw->closed`?

chu11 replied:
Yeah, that's what I was thinking, that the close callback could be called multiple times.  This function returns void, so should the errno matter?

chu11 commented on src/common/libflux/ev_buffer_write.c:
```diff
@@ -40,7 +42,16 @@ static void buffer_write_cb (struct ev_loop *loop, ev_io *iow, int revents)
 
-        if (!flux_buffer_bytes (ebw->fb))
+        if (!flux_buffer_bytes (ebw->fb) && ebw->eof) {
+            if (close (ebw->eof) < 0)
```
should be ```close (ebw->fd)```?

grondo replied:
Yipes, what a ridiculously ridiculous typo. Funny thing is in my `reactorcat` test this actually worked because it resolved to `close (1)` or `close (STDOUT_FILENO)`. (Luckily would be caught by the unit tests, but nice catch @chu11!)

chu11 commented:
Chatted with @grondo in the hall, after this PR I will add a "I'm finished with this buffer" flag into ```flux_buffer_t``` to handle the issue of  ```If data is written to the underlying flux_buffer_t after this call, the results are undefined.```

grondo commented:
@chu11, if you want to add the `flux_buffer_close()` call we could merge that PR first and then it could be utilized in this PR. (or we could do it as one PR)

chu11 commented:
I was just thinking it would be a separate PR, but lemme see if I can whip something up quick.

grondo commented:
> I was just thinking it would be a separate PR, but lemme see if I can whip something up quick.

Up to you but it would be nice to have the feature so it could be used in this PR. If you submit a PR I'll merge it and then use it in this one.

I still have to write tests here (and make the fixes you caught), so you have some time ;-)

grondo commented:
@chu11, pushed a commit to address your comments. `flux_buffer_write_watcher_close()` now returns and integer and indicates failure for these cases:

 * Invalid watcher handle: `EINVAL`
 * eof already set: `EINPROGRESS`
 * closed: `EINVAL`

Though it likely won't be useful to differentiate, it seemed simple to do and will aid in testing.

Also fixed the typo in the call the `close` :man_facepalming: 

Now to write some tests.

grondo commented:
Just added a small change to mark buffer read-only when `flux_buffer_write_watcher_close()` is called, and a couple tests to ensure this new code after close.

I also added 1 small test to `t0001-basic.t` to ensure the reactor-based cat test works to shuttle data from input to output and that EOF before any data (e.g. reading from `/dev/null`) also works as expected.

chu11 commented on src/common/libflux/reactor.h:
```diff
@@ -91,13 +91,27 @@ flux_watcher_t *flux_buffer_read_watcher_create (flux_reactor_t *r, int fd,
 
+/* "write" EOF to buffer write watcher 'w'. The underlying fd will be closed
+ *  once the buffer is emptied. If data is written to the underlying
+ *  flux_buffer_t after this call, the results are undefined.
```
"If data is written to the underlying flux_buffer_t after this call, the results are undefined." can probably be removed now.  Or if we want to, we can say attempts to write to the buffer result in EROFS.

grondo replied:
Good catch, thanks!

chu11 commented on src/common/libflux/test/reactor.c:
```diff
@@ -577,6 +579,59 @@ static void test_buffer (flux_reactor_t *reactor)
+    ok (flux_buffer_write (fb, "shouldfail", 10) == -1 && errno == EROFS,
+        "buffer: flux_buffer_write after close fails with EROFS");
+
+    flux_watcher_start (w);
```
should probably add a ```flux_buffer_write_watcher_is_closed() == 0``` test case before the reactor run?

grondo replied:
Good suggestion, will do.

grondo commented:
Ok, @chu11, added a couple quick commits to address your comments.

chu11 commented:
everything LGTM, lets squash, then merge it

grondo commented:
Thanks @chu11, squashed and force-pushed.

grondo commented:
Not sure what was going on, but Travis didn't initiate tests on the latest version of this PR. I've just force pushed again to wake it up.

grondo commented:
Thanks!
